bundles/org.openhab.binding.livisismarthome/src/main/java/org/openhab/binding/livisismarthome/internal/handler/LivisiBridgeHandler.java

```diff
@@ -0,0 +1,1004 @@
+            }
+        } catch (IOException | ApiException | AuthenticationException | RuntimeException e) {
+            logger.debug("Error with Event: {}", e.getMessage(), e);
+            handleClientException(e);
+        }
+    }
+
+    @Override
+    public void onError(final Throwable cause) {
+        if (cause instanceof Exception) {
+            handleClientException((Exception) cause);
+        }
+    }
+
+    /**
+     * Handles the event that occurs, when the state of a device (like reachability) or a capability (like a temperature
+     * value) has changed.
+     *
+     * @param event event
+     */
+    private void handleStateChangedEvent(final EventDTO event, @Nullable final DeviceStructureManager deviceStructMan)
+            throws ApiException, IOException, AuthenticationException {
+
+        if (deviceStructMan != null) {
+            // CAPABILITY
+            if (event.isLinkedtoCapability()) {
+                logger.trace("Event is linked to capability");
+                final Optional<DeviceDTO> device = deviceStructMan.getDeviceByCapabilityId(event.getSourceId());
+                notifyDeviceStatusListeners(device, event);
+
+                // DEVICE
+            } else if (event.isLinkedtoDevice()) {
+                logger.trace("Event is linked to device");
+                final String sourceId = event.getSourceId();
+
+                final Optional<DeviceDTO> bridgeDevice = deviceStructMan.getBridgeDevice();
+                final Optional<DeviceDTO> device;
+                if (bridgeDevice.isPresent() && !sourceId.equals(bridgeDevice.get().getId())) {
+                    device = deviceStructMan.refreshDevice(sourceId, isSHCClassic());
+                } else {
+                    device = deviceStructMan.getDeviceById(sourceId);
+                }
+                notifyDeviceStatusListeners(device, event);
+
+            } else {
+                logger.debug("link type {} not supported (yet?)", event.getSourceLinkType());
+            }
+        }
+    }
+
+    /**
+     * Handles the event that occurs, when the connectivity of the bridge has changed.
+     *
+     * @param event event
+     */
+    private void handleControllerConnectivityChangedEvent(final EventDTO event,
+            @Nullable final DeviceStructureManager deviceStructMan)
+            throws ApiException, IOException, AuthenticationException {
+
+        if (deviceStructMan != null) {
+            final Boolean connected = event.getIsConnected();
+            if (connected != null) {
+                final ThingStatus thingStatus = createThingStatus(connected);
+                logger.debug("SmartHome Controller connectivity changed to {}.", thingStatus);
+                if (connected) {
+                    deviceStructMan.refreshDevices();
+                }
+                updateStatus(thingStatus);
+            } else {
+                logger.warn("isConnected property missing in event! (returned null)");
```

Comment:
I don't know what you mean here / what I should change. The event is guaranteed to be of type /event/ControllerConnectivityChanged, because this is checked before. The method is only executed when the type is /event/ControllerConnectivityChanged, see the onEvent method.

Reply:
Message is logged when this method returns `null`:
```
    public Boolean getIsConnected() {
        if (!isControllerConnectivityChangedEvent()) {
            return null;
        }
        return getProperties().getIsConnected();
    }
```

which means it returns `null` if this returns `false`:

```
    public boolean isControllerConnectivityChangedEvent() {
        return TYPE_CONTROLLER_CONNECTIVITY_CHANGED.equals(getType());
    }
```

And type is "/event/ControllerConnectivityChanged".

So it logs if type doesn't match. Therefore suggestion would be to change message "isConnected property missing in event! (returned null)" to something more explicit.

If it isn't expected to happen, instead of returning null you could throw a type of `RuntimeException` with appropriate message. This would be caught by `onEvent`, so you could drop the null check and logging within `handleControllerConnectivityChangedEvent`. The logic within `handleClientException` would then also be triggered, I don't know if it should or not.

Reply:
It is impossible that the type isn't matching (at this place).

onEvent has a switch-case which calls handleControllerConnectivityChangedEvent only when the type is /event/ControllerConnectivityChanged. And the getIsConnected method is only called from this method. Therefore there can't be another call with another type.

I have now removed the type check within EventDTO#getIsConnected(), because it is only called when the type matches and if not (sometime after possible code changes in the future), it would not be a problem, because the value would be NULL. Therefore a type check wouldn't be required, it would lead to the same result (NULL).